### PR TITLE
FE-452 Add datepicker and variant fields to A/B testing

### PR DESCRIPTION
**Pull request by jonambas**
- Uses redux-form's [FieldArray](https://redux-form.com/7.4.2/docs/api/fieldarray.md/) to generate variants.

### CODE REVIEW

File: src/components/typeahead/TemplateTypeahead.js [hunk 1]
```diff
@@ -11,7 +11,14 @@ export class TemplateTypeahead extends Component {
   };
 
   componentDidMount() {
-    this.props.listTemplates();
+    const { results, listTemplates } = this.props;
+
+    // For redux-form FieldArrays
+    // Subsequent GETs on this list will refresh any selector using it,
+    // which also forces a form reinitialize if used in a form's initialValues.
```
Comment (jonambas) [inline src/components/typeahead/TemplateTypeahead.js]: Might be a good idea to force `listTemplates` at the page level instead

File: src/pages/abTesting/components/AbTestCreateForm.js [hunk 2]
```diff
@@ -85,9 +87,12 @@ export class AbTestCreateForm extends Component {
 }
 
 function mapStateToProps(state) {
+  const selector = formValueSelector(formName);
+
   return {
     initialValues: {},
-    hasSubaccounts: hasSubaccountsSelector(state)
+    hasSubaccounts: hasSubaccountsSelector(state),
+    subaccountId: selector(state, 'subaccount.id') || 0
```
Comment (rajumsys) [inline src/pages/abTesting/components/AbTestCreateForm.js]: why is this needed?  for initial load, we can probably just set in initialValues.

File: src/pages/abTesting/components/StatusPanel.js [hunk 3]
```diff
@@ -16,7 +16,7 @@ const VersionSelector = ({ current, latest, id, subaccountId }) => {
     content: `View Version ${i + 1}`,
     to: `/ab-testing/${id}/${i + 1}${setSubaccountQuery(subaccountId)}`,
     component: Link,
-    visible: i + 1 !== current
+    visible: i + 1 !== Number(current)
```
Comment (rajumsys) [inline src/pages/abTesting/components/StatusPanel.js]: is this casting necessary?

File: src/pages/abTesting/components/fields/VariantsFields.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+            component={TemplateTypeaheadWrapper}
+            label='Select a default template'
+            placeholder='Type to search'
+            disabled={disabled}
+            subaccountId={subaccountId}
+          />
+        </Grid.Column>
+        <Grid.Column>
+          {
+            formValues.audience_selection === 'sample_size'
```
Comment (rajumsys) [inline src/pages/abTesting/components/fields/VariantsFields.js]: i think it'll be better readable if you put the final component name in a variable at top and here just render it; just like `MyRoute` in App.js

File: src/pages/abTesting/components/fields/VariantsFields.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    </Panel>
+    <FieldArray
+      name='variants'
+      component={RenderVariants}
+      formValues={formValues}
+      disabled={disabled}
+      subaccountId={subaccountId}
+    />
+
+    {/* This is a temporary hack to make sure at least some of the last typeahead is visible on screen */}
```
Comment (rajumsys) [inline src/pages/abTesting/components/fields/VariantsFields.js]: what is the permanent solution? :D

File: src/selectors/abTesting.js [hunk 6]
```diff
@@ -34,11 +37,21 @@ export const selectLatestVersionNumberFromParams = createSelector(
...
+  [selectAbTestFromParams, selectTemplates],
+  (test, templates) => {
 
-    // Strip everything that is not editable
-    const values = _.omit(test, ['created_at', 'updated_at', 'id', 'status', 'version']);
+    // Strip everything that is not editable, or what will be reshaped
+    const values = _.omit(test, ['created_at', 'updated_at', 'id', 'status', 'version', 'start_time', 'end_time', 'variants', 'default_template' ]);
+
+    // Find template objects of the default template and all variants
+    const default_template = findTemplateObject(templates, test.default_template);
```
Comment (rajumsys) [inline src/selectors/abTesting.js]: `defaultTemplate`?

File: src/pages/abTesting/components/fields/StatusFields.js [hunk 7]
```diff
@@ -14,14 +15,23 @@ const StatusFields = ({ disabled }) => (
...
+        name='dates'
+        left
+        showPresets={false}
+        textFieldProps={{
+          helpText: 'A test may run for a maximum of 30 days',
+          label: 'When should we run this test?',
+          prefix: <AccessTime />
+        }}
+        datePickerProps={{
+          disabledDays: { before: new Date() },
```
Comment (rajumsys) [inline src/pages/abTesting/components/fields/StatusFields.js]: we should probably fix datepicker (sometime later) to hide left arrow when we don't allow a previous date (than what is displayed). currently it shows the arrow but noop.

File: src/pages/abTesting/components/fields/VariantsFields.js [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+  <Field
+    name={`${namespace}.sample_size`}
+    label='Number of messages'
+    component={TextFieldWrapper} {...props} />
+);
+
+/*
+  If you're looking at this, refer to https://redux-form.com/7.4.2/examples/fieldarrays/
+ */
+export const RenderVariants = ({ fields, formValues, disabled, subaccountId }) => (
```
Comment (rajumsys) [inline src/pages/abTesting/components/fields/VariantsFields.js]: where are these, specifically `fields` coming from?

File: src/pages/abTesting/components/StatusPanel.js (hunk 3 above)
Comment (jonambas) [reply src/pages/abTesting/components/StatusPanel.js]: Yeah, `current` comes from `version` from router params, which is a string. I could cast in the selector instead maybe

File: src/pages/abTesting/components/fields/VariantsFields.js (hunk 5 above)
Comment (jonambas) [reply src/pages/abTesting/components/fields/VariantsFields.js]: Idk, making Popover respond to its position (like tooltips)

File: src/pages/abTesting/components/fields/StatusFields.js (hunk 7 above)
Comment (jonambas) [reply src/pages/abTesting/components/fields/StatusFields.js]: Ah yeah, it looks like it does the same thing on the reports too

File: src/pages/abTesting/components/fields/VariantsFields.js (hunk 8 above)
Comment (jonambas) [reply src/pages/abTesting/components/fields/VariantsFields.js]: From `FieldArray` - see https://redux-form.com/7.4.2/docs/api/fieldarray.md/
It's a synthetic array that redux-form maintains

File: src/pages/abTesting/components/AbTestCreateForm.js (hunk 2 above)
Comment (jonambas) [reply src/pages/abTesting/components/AbTestCreateForm.js]: There are no initialValues here as its the create form - `subaccountId` here is used only the typeahead, so it knows which templates to filter/show and for validation in another ticket. I can add a comment to make that clear

File: src/selectors/abTesting.js (hunk 6 above)
Comment (jonambas) [reply src/selectors/abTesting.js]: Think I should camel case? I'm not camel casing any of the other keys 🤔

Comment (rajumsys) [reply src/selectors/abTesting.js]: no, i meant the local variable only.

File: src/pages/abTesting/components/StatusPanel.js (hunk 3 above)
Comment (jonambas) [reply src/pages/abTesting/components/StatusPanel.js]: Found a bug while investigating this - `cDU` doesn't correctly reGET ab test when changing version

File: src/pages/abTesting/components/AbTestCreateForm.js (hunk 2 above)
Comment (rajumsys) [reply src/pages/abTesting/components/AbTestCreateForm.js]: i think i see why you did it.
